src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,445 @@
+
+let rec trace_concat_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (ensures (
+      tr <++> Nil == tr /\
+      Nil <++> tr == tr
+    ))
+    [SMTPatOr [[SMTPat (tr <++> Nil)]; [SMTPat (Nil <++> tr)]]]
+  = match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_nil hd
+
+let rec trace_concat_assoc (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (ensures ((tr1 <++> tr2) <++> tr3) == (tr1 <++> (tr2 <++> tr3)))
+  = match tr3 with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_assoc tr1 tr2 hd
+
+let rec trace_subtract_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (ensures (tr <--> Nil) == tr)
+    [SMTPat (tr <--> Nil)]
+  = match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_subtract_nil hd
+  // Proves with () with initial trace subtract def
+  // TODO: Remove comment when trace subtraction is resolved
+
+let trace_subtract_snoc_left (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures ((Snoc tr2 e) <--> tr1) == Snoc (tr2 <--> tr1) e)
+  = ()
+
+let trace_subtract_snoc_left' (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires Snoc? tr2 /\ (let Snoc hd _ = tr2 in tr1 <$ hd))
+    (ensures (
+      let Snoc hd e = tr2 in
+      (tr2 <--> tr1) == Snoc (hd <--> tr1) e
+    ))
+  = ()
+
+let trace_concat_snoc_right (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (ensures tr1 <++> (Snoc tr2 e) == Snoc (tr1 <++> tr2) e)
+  = ()
+
+let snoc_is_concat_singleton (#label_t:Type) (tr :trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (ensures Snoc tr e == (tr <++> (Snoc Nil e)))
+  = ()
+
+let rec trace_concat_grows (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures tr1 <$ (tr2 <++> tr3))
+    [SMTPat (tr1 <$ (tr2 <++> tr3)); SMTPat (tr1 <$ tr2)]
+  = match tr3 with
+    | Nil -> ()
+    | Snoc hd e -> trace_concat_grows tr1 tr2 hd
+
+
+/// Properties connecting trace concatenation, splitting, and subtraction.
+
+let rec trace_split_concat (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr})
+  : Lemma
+    (ensures (
+      let (tr1, e, tr2) = trace_split tr ts in
+      (Snoc tr1 e) <++> tr2 == tr
+    ))
+  = let Snoc hd e = tr in
+    if ts = last_ts tr then ()
+    else trace_split_concat hd ts
+
+let rec trace_subtract_matches_split (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2 /\ Snoc? tr1)
+    (ensures (
+      let (_, _, tl) = trace_split tr2 (length tr1 - 1) in
+      tl == (tr2 <--> tr1)
+    ))
+  = let Snoc hd e = tr2 in
+    let (_, _, tl) = trace_split tr2 (length tr1 - 1) in
+    grows_cases tr1 tr2;
+    eliminate tr1 == tr2 \/ tr1 <$ hd
+    returns tl == (tr2 <--> tr1)
+    with _. ()
+    and _. trace_subtract_matches_split tr1 hd
+
+let rec trace_subtract_concat_right (#label_t:Type) (tr1 tr2:trace_ label_t)
```

Comment:
What about naming it `trace_subtract_concat`, as we first subtract and then concat?